azure/services/agentpools/spec.go

```diff
@@ -331,110 +254,37 @@ func (s *AgentPoolSpec) Parameters(ctx context.Context, existing interface{}) (p
 				NetNetfilterNfConntrackBuckets: s.LinuxOSConfig.Sysctls.NetNetfilterNfConntrackBuckets,
 				NetNetfilterNfConntrackMax:     s.LinuxOSConfig.Sysctls.NetNetfilterNfConntrackMax,
-				VMMaxMapCount:                  s.LinuxOSConfig.Sysctls.VMMaxMapCount,
-				VMSwappiness:                   s.LinuxOSConfig.Sysctls.VMSwappiness,
-				VMVfsCachePressure:             s.LinuxOSConfig.Sysctls.VMVfsCachePressure,
+				VmMaxMapCount:                  s.LinuxOSConfig.Sysctls.VMMaxMapCount,
+				VmSwappiness:                   s.LinuxOSConfig.Sysctls.VMSwappiness,
+				VmVfsCachePressure:             s.LinuxOSConfig.Sysctls.VMVfsCachePressure,
 			}
 		}
 	}
 
-	agentPool := armcontainerservice.AgentPool{
-		Properties: &armcontainerservice.ManagedClusterAgentPoolProfileProperties{
-			AvailabilityZones:    availabilityZones,
-			Count:                &s.Replicas,
-			EnableAutoScaling:    ptr.To(s.EnableAutoScaling),
-			EnableUltraSSD:       s.EnableUltraSSD,
-			KubeletConfig:        kubeletConfig,
-			KubeletDiskType:      azure.AliasOrNil[armcontainerservice.KubeletDiskType]((*string)(s.KubeletDiskType)),
-			MaxCount:             s.MaxCount,
-			MaxPods:              s.MaxPods,
-			MinCount:             s.MinCount,
-			Mode:                 ptr.To(armcontainerservice.AgentPoolMode(s.Mode)),
-			NodeLabels:           nodeLabels,
-			NodeTaints:           nodeTaints,
-			OrchestratorVersion:  s.Version,
-			OSDiskSizeGB:         &s.OSDiskSizeGB,
-			OSDiskType:           azure.AliasOrNil[armcontainerservice.OSDiskType](s.OsDiskType),
-			OSType:               azure.AliasOrNil[armcontainerservice.OSType](s.OSType),
-			ScaleSetPriority:     azure.AliasOrNil[armcontainerservice.ScaleSetPriority](s.ScaleSetPriority),
-			ScaleDownMode:        azure.AliasOrNil[armcontainerservice.ScaleDownMode](s.ScaleDownMode),
-			SpotMaxPrice:         spotMaxPrice,
-			Type:                 ptr.To(armcontainerservice.AgentPoolTypeVirtualMachineScaleSets),
-			VMSize:               sku,
-			VnetSubnetID:         vnetSubnetID,
-			EnableNodePublicIP:   s.EnableNodePublicIP,
-			NodePublicIPPrefixID: s.NodePublicIPPrefixID,
-			Tags:                 tags,
-			EnableFIPS:           s.EnableFIPS,
-			LinuxOSConfig:        linuxOSConfig,
-		},
+	// When autoscaling is set, the count of the nodes differ based on the autoscaler and should not depend on the
+	// count present in MachinePool or AzureManagedMachinePool, hence we should not make an update API call based
+	// on difference in count.
+	if s.EnableAutoScaling && agentPool.Status.Count != nil {
+		spec.Count = agentPool.Status.Count
 	}
 
 	return agentPool, nil
 }
 
-// mergeSystemNodeLabels appends any kubernetes.azure.com-prefixed labels from the AKS label set
-// into the local capz label set.
-func mergeSystemNodeLabels(capzLabels, aksLabels map[string]*string) map[string]*string {
-	// if both are nil, return nil for no change
```

Comment:
Until ASO v2.4 is available with https://github.com/Azure/azure-service-operator/pull/3407, the "set to nil to delete all labels" behavior enabled by #4122 will instead go back to the "setting to nil doesn't change" which it did before.

Reply:
is that a regression? what's the eta for 2.4?

Reply:
It is a regression, but if ASO v2.4.0 is included for CAPZ v1.12 then users shouldn't notice. @matthchr Do you think the next ASO minor release will be cut before mid-November?

Reply:
ASO v2.4 ETA is next week, which should be enough time for us.